### PR TITLE
Powers in terms of logarithms

### PR DESCRIPTION
**Pull request by i-redbyte**
Powers in terms of logarithms

### CODE REVIEW

**Review by raklaptudirm — DISMISSED** [review]

File: math/power/powvialogarithm.go
Comment (tjgurwara99) [inline math/power/powvialogarithm.go:14]: shouldn't you keep the arguments `float64`? The `math.Log` and `math.Exp` use float64 so all the type casting you use will not be necessary.

```suggestion
func PowUseLog(a float64, b float64) float64 {
```

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]
Please consider the following changes. Overall, as a mathematician I really like this approach :)

Also, I think this will be faster too from the other approach even though the expensive operations are being used. These expensive operations are assembly commands in most modern architecture which go takes advantage of. If it is possible could you add a benchmark for these `power` algorithms? If you can't, then just consider the recommended changes and we can approve this PR. Thanks!

**Review by siriak — DISMISSED** [review]
LGTM apart from the one comment from @tjgurwara99

File: math/power/powvialogarithm.go
Comment (i-redbyte) [reply math/power/powvialogarithm.go]: Done

**Review by raklaptudirm — APPROVED** [review]

File: math/pi/spigotpi.go
Comment (tjgurwara99) [inline math/pi/spigotpi.go:49]: Wouldn't it be better to convert this to a simple switch case distinction?

File: math/pi/spigotpi.go
Comment (tjgurwara99) [inline math/pi/spigotpi.go:57]: Do we really need this?

Comment (tjgurwara99) [reply math/pi/spigotpi.go:57]: On the other hand it makes this more readable - keep it there. It's fine.

File: math/pi/spigotpi_test.go
Comment (tjgurwara99) [inline math/pi/spigotpi_test.go:22]: Can you make this local inside the `TestPiSpigot` function or create a factory function that generates these test cases? Without a factory or local initialisation it is possible to have conflict with other implementations. I'm aware that this is the only algorithm inside this directory but if someone adds another algorithm, it could create conflict - better to be safe then sorry 😄

File: math/power/powvialogarithm.go
Comment (tjgurwara99) [inline math/power/powvialogarithm.go:14]: Can you change the name of the function? if someone was to use this in there code to try it out they would have to use it like this:

```Go
import github.com/TheAlgorithms/Go/math/power
...

func main() {
    ...
    power.PowUseLog(a, b)
    ...
}

```

which is arguably not very reader friendly - specially if someone is trying to learn from - the main purpose of this repo 😄 

My recommendation would be that you have something like this:

```Go
import github.com/TheAlgorithms/Go/math/power
...

func main() {
    ...
    power.UsingLog(a, b)
    ...
}
```

so it would be easier to parse through - sorry this is actually quite pedantic but believe me, readability makes a huge difference when parsing through code 😄 .

```suggestion
func UsingLog(a float64, b float64) float64 {
```

File: math/power/powvialogarithm_test.go
Comment (tjgurwara99) [inline math/power/powvialogarithm_test.go:10]: Same as previous comment - could you move this to a local declaration/initialisation or factory generation instead? Same point as other test comment applies 😄

File: math/power/powvialogarithm_test.go
Comment (tjgurwara99) [inline math/power/powvialogarithm_test.go:28]: reflect the names if you change the name of the PowUseLog function

File: math/power/powvialogarithm_test.go
Comment (tjgurwara99) [inline math/power/powvialogarithm_test.go:40]: Same for this!

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]
Nice implementations! As a mathematician I really liked your implementations. Please consider the changes that I have requested.

Also, could I ask you to make new PR's if you are submitting more than one implementation? This makes it easier to track changes and doesn't put a lot of load on maintainers 😄 If submitting another PR is not possible, then just update the PR's description and title to reflect the changes you have made.

If you have any troubles understanding the material, let me know and I'll be happy to discuss.

Thanks for contributing and happy coding!

File: math/pi/spigotpi.go
Comment (tjgurwara99) [inline math/pi/spigotpi.go:12]: Could you also change the name of this function - `pi` is already in the package name so `Spigot` should suffice.

File: math/pi/spigotpi.go
Comment (tjgurwara99) [inline math/pi/spigotpi.go:19]: `digitsHeld` instead of `heldDigits`
```suggestion
	digitsHeld := 0
```

File: math/pi/spigotpi.go
Comment (tjgurwara99) [inline math/pi/spigotpi.go:27]: typo here
```suggestion
			remainder[j] = sum % (j*2 + 1)
```

File: math/pi/spigotpi.go
Comment (i-redbyte) [reply math/pi/spigotpi.go:43]: Done

File: math/pi/spigotpi.go
Comment (i-redbyte) [reply math/pi/spigotpi.go:19]: Done

File: math/pi/spigotpi.go
Comment (tjgurwara99) [reply math/pi/spigotpi.go:27]: Don't worry - I'm also a non-native speaker 😄

File: math/power/powvialogarithm.go
Comment (i-redbyte) [reply math/power/powvialogarithm.go:14]: Totally agree with you. Fixed names

File: math/power/powvialogarithm_test.go
Comment (i-redbyte) [reply math/power/powvialogarithm_test.go:28]: Done

File: math/power/powvialogarithm_test.go
Comment (i-redbyte) [reply math/power/powvialogarithm_test.go:10]: Done

File: math/pi/spigotpi.go
Comment (i-redbyte) [reply math/pi/spigotpi.go:49]: Done

File: math/pi/spigotpi.go
Comment (tjgurwara99) [reply math/pi/spigotpi.go:19]: is it resolved? - I don't see it outdated. I don't think you have changed this variable - you might have missed it, haha 😄

File: math/pi/spigotpi.go
Comment (tjgurwara99) [inline math/pi/spigotpi.go:15]: You are missing an `a` in the middle 😃 

```suggestion
	remainders := make([]int, boxes)
```

File: math/pi/spigotpi.go
Comment (tjgurwara99) [inline math/pi/spigotpi.go:44]: The string conversion is also something that can be handled inside the delChar function - sorry I missed it last time.

```suggestion
				pi = delChar(pi, i-k)
```

File: math/pi/spigotpi.go
Comment (tjgurwara99) [inline math/pi/spigotpi.go:58]: Just do

```suggestion
	return string(append(tmp[0:index], tmp[index+1:]...))
```

File: math/power/powvialogarithm.go
Comment (tjgurwara99) [inline math/power/powvialogarithm.go:17]: Good use of bitwise operations. Nice!!

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]
Everything looks good other than some minor typo's. Very good, keep it up!

File: math/pi/spigotpi.go
Comment (i-redbyte) [reply math/pi/spigotpi.go:58]: maybe then it's better like this:
func delChar(s string, index int) string{
	tmp := []rune(s)
	return string(append(tmp[0:index], tmp[index+1:]...))
}

File: math/pi/spigotpi.go
Comment (i-redbyte) [reply math/pi/spigotpi.go:15]: Done

File: math/pi/spigotpi.go
Comment (i-redbyte) [reply math/pi/spigotpi.go]: Done

File: math/pi/spigotpi.go
Comment (i-redbyte) [reply math/pi/spigotpi.go]: Done

**Review by tjgurwara99 — DISMISSED** [review]
All good except the `helfDigits` slice name. It should be `digitsHeld`. I'm approving now, but if you change it, I'll approve it again. Thanks for contributing.

**Review by tjgurwara99 — APPROVED** [review]
LGTM! Thanks for contributing!

**Pull request merged**
